### PR TITLE
Vagrant: Support DHCP

### LINKED ISSUE #873: Multiple Trellis Projects at once
## Feature Request

Currently, working on multiple Trellis projects at the same time is tricky due to Vagrant ip's clashing. Wonder if something like this would be a good addition to the stack?

https://github.com/oscar-stack/vagrant-auto_network

Cheers

### PR DESCRIPTION
**Pull request by TangRufus**
Allow running multiple Trellis machines at the same time.

Tested with 2 DHCP machines and 1 static IP machine running at the same time. Results:
```bash
$ cat /etc/exports
# VAGRANT-BEGIN: 501 78f3bfb6-4ae0-4305-b4a3-8cd68040aa6b
"/Users/me/Code/siteA/bedrock" 172.28.128.3 -alldirs -mapall=501:20
"/Users/me/Code/siteA/trellis" 172.28.128.3 -alldirs -mapall=501:20
# VAGRANT-END: 501 78f3bfb6-4ae0-4305-b4a3-8cd68040aa6b
# VAGRANT-BEGIN: 501 2bf7c671-523e-4ca5-9a81-e85797428609
"/Users/me/Code/siteB/bedrock" 172.28.128.4 -alldirs -mapall=501:20
"/Users/me/Code/siteB/trellis" 172.28.128.4 -alldirs -mapall=501:20
# VAGRANT-END: 501 2bf7c671-523e-4ca5-9a81-e85797428609
# VAGRANT-BEGIN: 501 8dafbb55-1263-4829-90e2-b76fb68323c0
"/Users/me/Code/siteC/site" 192.168.50.5 -alldirs -mapall=501:20
"/Users/me/Code/siteC/trellis" 192.168.50.5 -alldirs -mapall=501:20
# VAGRANT-END: 501 8dafbb55-1263-4829-90e2-b76fb68323c0

$ cat /etc/hosts
...
## vagrant-hostmanager-start id: 95b74937-d48d-468e-850a-ce2664ad1ea7
172.28.128.3	www.typist.dev
172.28.128.3	typist.dev
## vagrant-hostmanager-end

## vagrant-hostmanager-start id: 3fa4d5c6-050d-4545-bc5c-e4db40ec516c
172.28.128.4	ww2.typist.dev
172.28.128.4	ww3.typist.dev
## vagrant-hostmanager-end

## vagrant-hostmanager-start id: 5695f53b-c08e-46be-9d1f-dd08d9cc2842
192.168.50.5	example.dev
192.168.50.5	www.example.dev
## vagrant-hostmanager-end
```

The `ip_resolver` proc is taken from https://github.com/devopsgroup-io/vagrant-hostmanager/issues/86#issuecomment-176572400
Related https://github.com/roots/trellis/issues/295#issuecomment-129458413
Closes #873

### CODE REVIEW

File: Vagrantfile [hunk 1]
```diff
@@ -32,7 +32,23 @@ Vagrant.configure('2') do |config|
   config.ssh.shell = %{bash -c 'BASH_ENV=/etc/profile exec bash'}
 
   # Required for NFS to work
-  config.vm.network :private_network, ip: vconfig.fetch('vagrant_ip'), hostsupdater: 'skip'
+  if vconfig.fetch('vagrant_ip') == 'dhcp'
+    config.vm.network :private_network, type: 'dhcp', hostsupdater: 'skip'
+
+    cached_addresses = {}
+    config.hostmanager.ip_resolver = proc do |vm, resolving_vm|
```
Comment (swalkinshaw) [inline Vagrantfile]: Small thing, but prefix `resolving_vm` with `_` since it's unused

Comment (TangRufus) [reply Vagrantfile]: Done

**Comment by swalkinshaw** [comment]
We may even want to make dhcp the default. Seems simpler?

**Comment by TangRufus** [comment]
👍  for default

However, if you already have a VM with static IP (running / poweroff doesn't matter), you need `$ vagrant destroy` before switching to DHCP.

**Comment by swalkinshaw** [comment]
I think just `vagrant reload` works?

**Comment by TangRufus** [comment]
I had trouble with `vagrant reload`. 
But, it could cause by `vagrant-triggers`, didn't investigate.

**Comment by oxyc** [comment]
> I had trouble with vagrant reload.

I tested this over at Drupal VM and same happened for me if you're referring to a `mount.nfs: access denied by server while mounting` error @TangRufus? `vagrant halt; vagrant up` fixed it though.

**Comment by TangRufus** [comment]
I got more errors besides `mount.nfs: access denied by server while mounting ...`. I guess they were from one of my vagrant plugins.

Just curious: Do you work while travelling, @oxyc? I found it extremely hard to stay in front of the computer when everyone else in the hostel went out for sun and beach.

**Comment by oxyc** [comment]
For reference the plugins I had were `vagrant-vbguest` and `vagrant-hostmanager`. Guess it might be triggers.

Nah I never liked mixing work and travel, wouldn't do either at full capacity. I usually travel 4-6 months until I get sick of it, then find a place to work for 6-8 months until I miss it again, and repeat. Also I like signing off so I don't want to bring my laptop.

**Pull request merged**